src/Concerns/Testing.php

```diff
@@ -78,44 +85,52 @@ trait Testing
      */
     final protected function tearDownTheTestEnvironment(): void
     {
-        if ($this->app) {
-            foreach ($this->beforeApplicationDestroyedCallbacks as $callback) {
-                \call_user_func($callback);
+        try {
+            if ($this->app) {
+                try {
+                    foreach ($this->beforeApplicationDestroyedCallbacks as $callback) {
+                        \call_user_func($callback);
+                    }
+                } finally {
+                    $this->app->flush();
+                    $this->app = null;
+
+                    foreach ($this->afterApplicationDestroyedCallbacks as $callback) {
+                        \call_user_func($callback);
+                    }
+                }
             }
+        } finally {
+            $this->setUpHasRun = false;
 
-            $this->app->flush();
+            if (\property_exists($this, 'serverVariables')) {
+                $this->serverVariables = [];
+            }
 
-            $this->app = null;
-        }
+            if (\property_exists($this, 'defaultHeaders')) {
+                $this->defaultHeaders = [];
+            }
 
-        $this->setUpHasRun = false;
+            if (\class_exists('Mockery')) {
+                if ($container = Mockery::getContainer()) {
+                    $this->addToAssertionCount($container->mockery_getExpectationCount());
+                }
 
-        if (\property_exists($this, 'serverVariables')) {
-            $this->serverVariables = [];
-        }
+                Mockery::close();
+            }
 
-        if (\property_exists($this, 'defaultHeaders')) {
-            $this->defaultHeaders = [];
-        }
+            Carbon::setTestNow();
 
-        if (\class_exists('Mockery')) {
-            if ($container = Mockery::getContainer()) {
-                $this->addToAssertionCount($container->mockery_getExpectationCount());
+            if (class_exists(CarbonImmutable::class)) {
+                CarbonImmutable::setTestNow();
             }
 
-            Mockery::close();
-        }
+            $this->afterApplicationCreatedCallbacks = [];
+            $this->beforeApplicationDestroyedCallbacks = [];
+            $this->afterApplicationDestroyedCallbacks = [];
 
-        Carbon::setTestNow();
-
-        if (class_exists(CarbonImmutable::class)) {
-            CarbonImmutable::setTestNow();
+            Artisan::forgetBootstrappers();
         }
-
-        $this->afterApplicationCreatedCallbacks = [];
-        $this->beforeApplicationDestroyedCallbacks = [];
-
-        Artisan::forgetBootstrappers();
```

Comment:
<img width="713" alt="Screen Shot 2019-07-22 at 9 33 21 AM" src="https://user-images.githubusercontent.com/172966/61600549-cd713900-ac63-11e9-96ff-1c167bffb421.png">

I don't get why we need two nested `try`, surely one is enough since outside of `if ($this->app)` condition we are only unsetting the values.

Reply:
The nested tries are from @GrahamCampbell he can correct me if I am wrong.

As far as I understand the idea is that if in the same tearDown 

- a beforeApplicationDestroyed callback throws the app will still get flushed.
- an afterApplicationDestroyedCallback throws the variables will still get reset.

My initial fix had only the outter try/finally but it makes the assumption that beforeApplicationDestroyed callbacks won't throw.

While being a bit clunky the nested try/finally seems to really add value here. Or maybe there is another way?

Reply:
I would just use `rescue()` for this TBH.

```php
if ($this->app) {
    foreach ($this->beforeApplicationDestroyedCallbacks as $callback) {
        \rescue($callback, null, false);
    }
    
    $this->app->flush();
    $this->app = null;

    foreach ($this->afterApplicationDestroyedCallbacks as $callback) {
        \rescue($callback, null, false);
    }
}
```

cc @GrahamCampbell

Reply:
Oh I didn't know about this helper.

Well then I guess using `rescue ` is definitely a better option.

I'll make the changes.

Reply:
Ahh actually we can't use `rescue` here.

calls to the `this->fails()` method of phpunit throws. If we use rescue then phpunit won't be able to display these errors.

I guess we can call the `beforeApplicationDestroyedCallbacks` with rescue and the afterApplicationDestroyedCallbacks without it.

If a test trait needs to call `$this->fail()` it will need to do so in an `afterApplicationDestroyed` callback. 

I'll check if laravel's provided test trait do so and update them accordingly if we go that route.

Maybe @GrahamCampbell as another idea?